titles: use rdrama's title finding code

justcool393 opened the pull request:
this fixes a potential DoS in some really weird pages (seems to be a bug with BS4) 
we're not parsing arbitrary HTML and the title tag is probably one of the simplest tags so i think this is acceptable
in addition we make some nice checks

justcool393 committed:
titles: use rdrama's title finding code
this fixes a potential DoS in some really weird pages (seems to be a bug with BS4)
we're not parsing arbitrary HTML
in addition we make some nice checks

justcool393 commented:
reference: https://github.com/Aevann1/rDrama/commit/f5912bb4cc0dd3d0c7a88b32f077255e3b6dd186 (and a couple of commits that came after it but i forgot what they were)

justcool393 commented:
![image](https://user-images.githubusercontent.com/1238820/203096926-1cd127b5-a921-4cb1-8b69-dab48fd9397e.png)

![image](https://rdrama.net/e/marseyagreefast.webp)

TLSM commented:
The precise nature of the DoS was that a dependency of BS4 would get caught in an infinite loop doing charset detection on certain pages (most frequently some YouTube pages). We still have no idea why, even after investigating the dependency code itself. However, it was sufficient to crash the worker and max out a core for quite some time.

justcool393 commented:
apparently looking for meta charset tags or meta http-equiv tags is too hard for BS4 :pensive:

justcool393 commented:
regardless this should be faster anyway since we don't actually need to parse the HTML nor even really care about its contents other than the title. requests parses the HTTP thing from Content-Type or Content-Encoding headers so it shouldn't be an issue

and if it isn't well formed enough for this it's not worth caring about anyway since there are probably other major issues with the page being linked to

TLSM commented:
Bug discovered on upstream, not yet integrated into this PR: need to run retrieved title through `html.unescape` before returning. Many sites include HTML character references in their `<title>`, which otherwise display literally when used as an `<input>` value—which is a safe context that doesn't need character references.

justcool393 committed:
unescape title to fix bug from upstream

justcool393 commented:
fixed @TLSM

justcool393 committed:
fix nameerror

justcool393 commented on files/routes/posts.py:
```diff
@@ -1147,16 +1147,24 @@
+	if not url.startswith('http'): abort(400)
+
+	checking_url = url.lower().split('?')[0].split('%3F')[0]
+	if any((checking_url.endswith(f'.{x}') for x in extensions)):
```
![image](https://rdrama.net/e/marseyagreefast.webp)

TLSM committed:
Do not proxy requests, since no proxy available.

On the upstream, the `proxies` dict was intended to use a local SOCKS
proxy running on port 18080 with the express purpose of masking the
server IP address. TheMotte isn't running behind a reverse proxy, so
this purpose is moot. Additionally, we don't have a proxy running in
Docker nor do we appear to have one on prod, which breaks autotitle
and thumbnailing regardless--not sure it matters for TheMotte's
use case, but both codepaths have been inoperative because of it.

TLSM reviewed (approved):
Faithful replica of operative and stable code on the upstream. Tested locally and works. Fixes an actual DoS.

justcool393 committed:
use gevent to timeout the function to prevent a
second theoretical DoS by sending data rly slowly
ref: https://fsdfsd.net/rDrama/rDrama/commit/816389cf28749bd055a9dc4af32538a9ecea5bc2
